### PR TITLE
Add foldCopyData helper function

### PR DESCRIPTION
**Pull request by sestrella**
This PR introduces a new function called foldCopyData which folds over `COPY TO STDOUT` query.

### CODE REVIEW

**Review by phadej — CHANGES_REQUESTED** [review]
The `withCopyData` and `getAllCopyData` are too specific functions.

I'd rather write something like

```haskell
-- | Fold over @COPY TO STDOUT@ query. ... same notes as for 'getCopyData'.
foldCopyData
    :: (a -> ByteString -> IO a)  -- ^ Accumulate one row of the result
    -> (a -> Int64 -> IO b)       -- ^ Post-process accumulator with a count of rows
    -> a                          -- ^ Initial accumulator
    -> IO b                       -- ^ Result
```

and let people specialise.

Your `getAllCopyData` has performance issue, `++` is quadratic when used like that.
Check difference lists (i.e. composing `[a] -> [a]` functions).
It would be a simple instantiation of `foldCopyData`

**Comment by sestrella** [comment]
@phadej thank you for the feedback! I completely agree with you, I think both functions could be replaced by a more general function.

File: src/Database/PostgreSQL/Simple/Copy.hs [hunk 1]
```diff
@@ -104,6 +105,30 @@ data CopyOutResult
...
+--   >     (\acc count -> return (acc, count))
+--   >     []
+
+foldCopyData
+  :: Connection                   -- ^ Database connection
+  -> (a -> B.ByteString -> IO a)  -- ^ Accumulate one row of the result
+  -> (a -> Int64 -> IO b)         -- ^ Post-process accumulator with a count of rows
+  -> a                            -- ^ Initial accumulator
+  -> IO b                         -- ^ Result
+foldCopyData conn f g acc = do
```
Comment (phadej) [inline src/Database/PostgreSQL/Simple/Copy.hs:125]: Make `acc` strict, i.e. `foldCopyData conn f g !acc = ...`

File: src/Database/PostgreSQL/Simple/Copy.hs [hunk 2]
```diff
@@ -104,6 +105,30 @@ data CopyOutResult
...
+    result <- getCopyData conn
+    case result of
+      CopyOutRow row    -> f acc row >>= foldCopyData conn f g
```
Comment (phadej) [inline src/Database/PostgreSQL/Simple/Copy.hs:128]: 4 spaces

**Review by phadej — COMMENTED** [review]
Looks better!

File: src/Database/PostgreSQL/Simple/Copy.hs (hunk 1 above)
Comment (sestrella) [reply src/Database/PostgreSQL/Simple/Copy.hs]: Done, I just made `acc` strict

File: src/Database/PostgreSQL/Simple/Copy.hs (hunk 2 above)
Comment (sestrella) [reply src/Database/PostgreSQL/Simple/Copy.hs]: My bad, I thought I covered all styling issues. I added some extra space.

**Comment by phadej** [comment]
Good, thanks!

**Comment by phadej** [comment]
Released in [`postgresql-0.6.4`](https://hackage.haskell.org/package/postgresql-simple-0.6.4)

**Pull request merged**